Fix dead code and string concatenation bugs in poisson_nll_loss

gaurav-redhat opened the pull request:
### Summary

This change fixes two small but real bugs in the error handling of `poisson_nll_loss` that could cause confusing behavior or crashes for users.

### What was fixed

**1. Removed dead code**

There was an unused variable assignment right before an error was raised. Since the value was never used, it served no purpose and was removed to keep the code clean and correct.

**2. Made error messages safer**

The error message used string concatenation, which could crash if `reduction` was not a string. This has been replaced with an f-string that works safely for all input types and clearly shows the valid options.

### What changed

* Removed unnecessary code that was never used
* Replaced unsafe string concatenation with a safer f-string
* Improved the error message by listing the valid `reduction` values
* Prevents accidental `TypeError` when users pass an invalid type

### Impact

* More robust and predictable error handling
* Clearer error messages for users
* Small functional fix with no behavior changes beyond better errors

gaurav-redhat committed:
Fix dead code and string concatenation bugs in poisson_nll_loss

COMPOUND CODING BUG FIX: Fixed two functional bugs in error handling:

1. DEAD CODE BUG: Removed pointless 'ret = input' assignment that was
   never used (error raised immediately after)

2. STRING CONCATENATION BUG: Replaced unsafe concatenation with f-string
   formatting to prevent TypeError crashes

- Eliminates dead code that confused readers
- Prevents runtime TypeError when reduction is not a string
- Provides clearer error messages with expected values
- Uses modern f-string formatting

This addresses multiple real functional bugs in the same function.

zou3519 commented on torch/nn/functional.py:
```diff
@@ -3237,8 +3237,7 @@ def poisson_nll_loss(
     if reduction != "none" and reduction != "mean" and reduction != "sum":
-        ret = input
-        raise ValueError(reduction + " is not a valid value for reduction")
+        raise ValueError(f"'{reduction}' is not a valid reduction mode. Expected 'none', 'mean', or 'sum'.")
```
please combine this PR with the last PR (https://github.com/pytorch/pytorch/pull/171077/files)

zou3519 reviewed (changes requested):
see comment

gaurav-redhat commented:
@pytorchbot label "topic: not user facing"

gaurav-redhat committed:
Add test for poisson_nll_loss invalid reduction error message

gaurav-redhat committed:
Remove unnecessary backslash escapes in test regex

Address review comment from isuruf: remove backslash escapes
before dots in assertRaisesRegex pattern string.

Made-with: Cursor
